doc/src/vpr/fasm.rst

```diff
@@ -0,0 +1,264 @@
+          BLK_IG-COMMON_SLICE.DX : D5FFMUX.IN_B
+        </meta>
+      </metadata>
+    </mux>
+
+The above mux connects input BLK_IG-COMMON_SLICE.DX or BLK_IG-COMMON_SLICE.DO5
+to BLK_BB-SLICE_FF.D5[3].  When VPR selects BLK_IG-COMMON_SLICE.DO5 for the
+mux, "genfasm" will emit D5FFMUX.IN_A, etc.
+
+There is not a requirement that all inputs result in a feature being set.
+In cases where some mux selections result in no feature being set, use "NULL"
+as the feature name.  Example:
+
+.. code-block:: xml
+
+    <mux name="CARRY_DI3" input="BLK_IG-COMMON_SLICE.DO5 BLK_IG-COMMON_SLICE.DX" output="BEL_BB-CARRY[2].DI" >
+      <metadata>
+        <meta name="fasm_mux">
+          BLK_IG-COMMON_SLICE.DO5 : CARRY4.DCY0
+          BLK_IG-COMMON_SLICE.DX : NULL
+        </meta>
+      </metadata>
+    </mux>
+
+The above examples all used the ``<mux>`` tag.  The "fasm_mux" metadata key
+can also be used with the ``<direct>`` tag in the same way, example:
+
+.. code-block:: xml
+
+    <direct name="WA7"  input="BLK_IG-SLICEM.CX" output="BLK_IG-SLICEM_MODES.WA7">
+      <metadata>
+        <meta name="fasm_mux">
+          BLK_IG-SLICEM.CX = WA7USED
+        </meta>
+      </metadata>
+    </direct>
+
+Passing parameters through to the FASM Output
+---------------------------------------------
+
+In many cases there are parameters that need to be passed directly from the
+eblif input to the FASM file.  These can be passed into a FASM feature via the
+"fasm_params" key.  Note that care must be taken to have the "fasm_params"
+metadata be attached to pb_type that the packer uses, the pb_type with the
+blif_model= ".subckt".
+
+The "fasm_params" value is a newline separated list of FASM features to eblif
+parameters.   Example:
+
+.. code-block:: xml
+
+  <metadata>
+    <meta name="fasm_params">
+      INIT[31:0] = INIT_00
+      INIT[63:32] = INIT_01
+    </meta>
+  </metadata>
+
+The FASM feature is on the left hand side of the equals.  When setting a
+parameter with multiple bits, the bit range must be specified.  If the
+parameter is a single bit, the bit range is not required, but can be supplied
+for clarity.  The right hand side is the parameter name from eblif.  If the
+parameter name is not found in the eblif, that FASM feature will not be
+emitted.
+
+No errors or warnings will be generated for unused parameters from eblif or
+unused mappings between eblif parameters and FASM parameters to allow for
+flexibility in the synthesis output.  This does mean it is important to check
+spelling of the metadata, and create tests that the mapping is working as
+expected.
+
+Also note that "genfasm" will not accept "x" (unknown/don't care) or "z"
+(high impedence) values in parameters.  Prior to emitting the eblif for place
+and route, ensure that all parameters that will be mapped to FASM have a
+valid "1" or "0".
+
+Invoking genfasm
+----------------
+
+``genfasm`` expects that place and route on the design is completed, so ensure
```

Comment:
This section should probably be near the top of the documentation (just after the general overview), so that those only interested in running genfasm can stop reading before the details of the internal implementation.

Reply:
Sure